client/client_test.go

```diff
@@ -2049,6 +2051,107 @@ func testBuildPushAndValidate(t *testing.T, sb integration.Sandbox) {
 	require.False(t, ok)
 }
 
+func testStargzLazyPull(t *testing.T, sb integration.Sandbox) {
+	skipDockerd(t, sb)
+	requiresLinux(t)
+
+	cdAddress := sb.ContainerdAddress()
+	if cdAddress == "" || !sb.Stargz() {
+		t.Skip("test requires containerd worker with stargz snapshotter")
+	}
+
+	client, err := newContainerd(cdAddress)
+	require.NoError(t, err)
+	defer client.Close()
+	registry, err := sb.NewRegistry()
+	if errors.Is(err, integration.ErrorRequirements) {
+		t.Skip(err.Error())
+	}
+	require.NoError(t, err)
+
+	// Prepare stargz image
+	sgzImage := registry + "/stargz/alpine:latest"
+	err = exec.Command("ctr-remote", "image", "optimize",
+		"--period=1", "alpine:latest", sgzImage).Run()
+	require.NoError(t, err)
+
+	c, err := New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	// stargz layers should be lazy even for executing something on them
+	def, err := llb.Image(sgzImage).
+		Run(llb.Args([]string{"true"})).
+		Marshal(context.TODO())
+	require.NoError(t, err)
+	target := registry + "/buildkit/testlazyimage:latest"
+	_, err = c.Solve(context.TODO(), def, SolveOpt{
+		Exports: []ExportEntry{
+			{
+				Type: ExporterImage,
+				Attrs: map[string]string{
+					"name": target,
+					"push": "true",
+				},
+			},
+		},
+	}, nil)
+	require.NoError(t, err)
+
+	var (
+		imageService = client.ImageService()
+		contentStore = client.ContentStore()
+		ctx          = namespaces.WithNamespace(context.Background(), "buildkit")
+	)
+
+	img, err := imageService.Get(ctx, target)
+	require.NoError(t, err)
+
+	manifest, err := images.Manifest(ctx, contentStore, img.Target, nil)
+	require.NoError(t, err)
+
+	// Check if image layers are lazy
+	var layers []ocispec.Descriptor
+	for _, layer := range manifest.Layers[:len(manifest.Layers)-1] {
```

Comment:
Thanks for the comment. This is because the topmost layer (created by running a command on that stargz image) isn't lazy. I added comments about this and also added a check to make sure that the topmost layer isn't lazy.

Reply:
Ohh duh, forgot that the topmost layer was created by run, makes sense, thanks